File: SCC/SCC.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <map>
#include <unordered_map>
#include <sstream>
#include <algorithm>

using Graph = std::map<int, std::vector<int>>;

std::ostream &operator<<(std::ostream &out, const std::map<int, int>& m) {
    for (auto e : m) out << '{' << e.first << ',' << e.second << "} ";
    return out;
}

std::ostream &operator<<(std::ostream &out, const std::vector<int> &v) {
    for (auto e : v) out << e << ' ';
    return out;
}

std::ostream &operator<<(std::ostream &out, const Graph &g) {
    for (auto t : g) out << t.first << ": " << t.second << std::endl;
    return out;
}

Graph readInput(std::string fileName) {
    std::ifstream in{ fileName };
    int src;
    Graph res;
    while (in >> src) {
        std::string line;
        std::getline(in, line);
        std::stringstream stream(line);
        int dst;
        while (stream >> dst) res[src].push_back(dst);
    }
    return res;
}

Graph reverseGraph(const Graph &in) {
    Graph out;
    for (const auto &it : in) {
        for (auto e : it.second)
            out[e].push_back(it.first);
    }
    return out;
}

constexpr auto magic = 900000;

std::vector<int> finishingTime(const Graph &g) {
    std::vector<int> ft(magic, -1);
    std::vector<bool> visited(magic, false);
    int timer = 0;
    std::vector<int> stack;
    stack.reserve(magic);

    auto dfs = [&] {
        while (!stack.empty()) {
            int cur = stack.back();
            stack.pop_back();

            if (cur < 0) {
                ft[timer++] = -cur;
                continue;
            }

            if (visited[cur]) continue;
            stack.push_back(-cur);

            visited[cur] = true;

            const auto &edges = g.find(cur);
            if (edges == g.end()) continue;
            for (auto it = edges->second.rbegin(); it != edges->second.rend(); ++it)
                if (!visited[*it]) stack.push_back(*it);
        }
    };

    for (auto it = g.rbegin(); it != g.rend(); ++it) {
        int node = it->first;
        if (visited[node]) continue;
        stack.push_back(node);
        dfs();
    }

    ft.resize(timer);
    return ft;
}

std::map<int, int> connectedComponents(const Graph &g, const std::vector<int> &ft) {
    std::vector<bool> visited(magic, false);
    std::map<int, int> componentCount;
    std::vector<int> stack;
    stack.reserve(magic);

    int leader = -1;

    auto dfs = [&] {
        while (!stack.empty()) {
            int cur = stack.back();
            stack.pop_back();

            if (visited[cur]) continue;

            ++componentCount[leader];
            visited[cur] = true;

            const auto &edges = g.find(cur);
            if (edges == g.end()) continue;
            for (auto it = edges->second.rbegin(); it != edges->second.rend(); ++it)
                if (!visited[*it]) stack.push_back(*it);
        }
    };

    for (auto it = ft.rbegin(); it != ft.rend(); ++it) {
        if (visited[*it]) continue;
        leader = *it;
        stack.push_back(*it);
        dfs();
    }
    return componentCount;
}

int main(int argc, char *argv[]) {
    if (argc < 2) return 1;
    const auto g = readInput(argv[1]);
    std::cout << g.size() << std::endl;

    const auto gr = reverseGraph(g);
    std::cout << gr.size() << std::endl;

    const auto ft = finishingTime(gr);
    std::cout << ft.size() << std::endl;

    const auto cc = connectedComponents(g, ft);
    std::cout << cc.size() << std::endl;

    std::vector<int> weights;
    for (auto it = cc.begin(); it != cc.end(); ++it)
        weights.push_back(it->second);
    std::sort(weights.rbegin(), weights.rend());

    std::cout << "weights ";
    for (int i = 0; i < std::min(weights.size(), 5UL); ++i)
        std::cout << weights[i] << ' ';
    std::cout << std::endl;
}

File: SCC/SCC.hs
import qualified Data.Map.Strict as Map
import qualified Data.Vector.Unboxed as V
import qualified Data.Vector.Unboxed.Mutable as MV
import System.Environment (getArgs)
import Control.Applicative (liftA2)
import Data.Either (fromRight)
import Control.Monad.ST
import Control.Monad
import Data.STRef
import qualified Data.Text as T
import qualified Data.Text.IO as TIO
import Data.Text.Read (decimal)
import Data.List (singleton, sortBy)

type Graph = Map.Map Int (V.Vector Int)

readInput :: String -> IO Graph
readInput file = do
    let parse = fst . fromRight (-1, T.empty) . decimal
    edges <- map (liftA2 (,) head (singleton . (!! 1)) . map parse . T.words) . T.lines <$> TIO.readFile file
    return $ (Map.map V.fromList . Map.fromListWith (flip (++))) edges

reverseGraph :: Graph -> Graph
reverseGraph g = runST $ do
    out <- newSTRef (Map.empty :: Map.Map Int [Int])
    forM_ (Map.assocs g) $ \(k, vs) -> do
        V.forM_ vs $ \v -> do
            isMember <- Map.member v <$> readSTRef out
            unless isMember $ modifySTRef out (Map.insert v [])
            modifySTRef out (Map.adjust (++ [k]) v)
    Map.map V.fromList <$> readSTRef out

doWhile :: Monad m => m Bool -> m a -> m ()
doWhile cond act = do
    res <- cond
    when res (act >> doWhile cond act)

finishingTime :: Graph -> V.Vector Int
finishingTime g = runST $ do
    let magic = 900000 :: Int
    ft <- MV.replicate magic (-1 :: Int)
    visited <- MV.replicate magic False
    timer <- newSTRef (0 :: Int)
    stack <- newSTRef []

    let dfs = doWhile (not . null <$> readSTRef stack) $ do
        cur <- head <$> readSTRef stack
        modifySTRef stack tail
        if cur < 0 then do
            now <- readSTRef timer
            modifySTRef timer (1 +)
            MV.write ft now (-cur)
        else do
            seen <- MV.read visited cur
            unless seen $ do
                modifySTRef stack (-cur :)
                MV.write visited cur True
                when (Map.member cur g) $ do
                    let edges = V.reverse $ g Map.! cur
                    V.forM_ edges $ \edge -> do
                        edgeSeen <- MV.read visited edge
                        unless edgeSeen (modifySTRef stack (edge :))

    forM_ (reverse $ Map.keys g) $ \node -> do
        seen <- MV.read visited node
        unless seen $ do
            modifySTRef stack (node :)
            dfs

    endTime <- readSTRef timer
    let slice = MV.take endTime ft
    V.freeze slice

connectedComponents :: Graph -> V.Vector Int -> Map.Map Int Int
connectedComponents g ft = runST $ do
    let magic = 900000 :: Int
    visited <- MV.replicate magic False
    stack <- newSTRef []
    leader <- newSTRef 0
    componentCount <- newSTRef (Map.empty :: Map.Map Int Int)

    let dfs = doWhile (not . null <$> readSTRef stack) $ do
        cur <- head <$> readSTRef stack
        modifySTRef stack tail
        seen <- MV.read visited cur
        unless seen $ do
            leader' <- readSTRef leader
            leaderMissing <- Map.notMember leader' <$> readSTRef componentCount
            when leaderMissing $ modifySTRef componentCount (Map.insert leader' 0)
            modifySTRef componentCount (Map.adjust (1 +) leader')
            MV.write visited cur True

            when (Map.member cur g) $ do
                V.forM_ (V.reverse $ g Map.! cur) $ \edge -> do
                    edgeSeen <- MV.read visited edge
                    unless edgeSeen (modifySTRef stack (edge :))

    V.forM_ (V.reverse ft) $ \node -> do
        seen <- MV.read visited node
        unless seen $ do
            modifySTRef leader (const node)
            modifySTRef stack (node :)
            dfs

    readSTRef componentCount


main :: IO ()
main = do
    input <- getArgs >>= readInput . head
    print $ Map.size input

    let reversed = reverseGraph input
    print $ Map.size reversed

    let ft = finishingTime reversed
    print $ V.length ft

    let cc = connectedComponents input ft
    print $ Map.size cc

    print $ (take 5 . sortBy (flip compare) . map snd . Map.assocs) cc
